src/V3LinkCells.cpp

```diff
@@ -88,6 +88,29 @@ void LinkCellsGraph::loopsMessageCb(V3GraphVertex* vertexp) {
     }
 }
 
+//######################################################################
+// Rename encoded names to pretty name
+
+class RestorePrettyNameVisitor : public AstNVisitor {
```

Comment:
I think this is a bit confusing, as then name() is sometimes encoded and sometimes pretty.

Note prettyName will result in text that might not be legal C++.  I presume that's ok everywhere you're then going to use name()?

Is there a good alternative?
1. Could prettyName be used instead where needed?
2. Could we avoid changing the name in the first place?
3. Track name and prettyName as one object so both are around?
4. other?

Reply:
I agree it's a bit tricky.

The trick is necessary when loading a protectlib-wrapper wiith parameter specialization.
Parser automatically encodes any symbols that contains `__` . (I understand that this encoding guarantees Verilator to create symbol without conflict)
Parser also encodes function names imported by "DPI-C"  then caused "undefined symbol" error.

I initially tried 2., but because the encoding happens in very early stage where everything is a token.
It was hard to distinguish which token to rename and which are not.

1. seems more torublesome than current renaming scheme because it will affect C++ code emitter.


Now I want to try 2 again.
If special metacomment or pragma is added to protect-lib wrapper, parser can stop encoding even when '__' is found.
I think the following change would be sufficient.
- protect-lib wrapper emitter to add extra pragma
- parser to stop encoding if the pragma is found

Any comment feedback is appreciated.

Reply:
Now I want to try 2 again.
If special metacomment or pragma is added to protect-lib wrapper, parser can stop encoding even when '__' is found.
I think the following change would be sufficient.
- protect-lib wrapper emitter to add extra pragma
- parser to stop encoding if the pragma is found

Any feedback is appreciated.

Reply:
While the pragma trick may work, basically the output verilated code then wouldn't work in another simulator that didn't understand the pragma, which is a high level goal of protect-lib.

As it does seem better to have the name be as close as possible to the original, that suggests track 1  - that is as output verilog should be as close as possible to input, when we write the wrapper it should use origName (or prettyName)?

Reply:
Let me explain more detail (hopefully in a better than before)
.
Assume the steps below.
Step1: Create a protect-lib by Verilator (Using -GParam=Var option to specialize parameterized top module)
Step2: Load protect-lib wrapper generated by Step1 and other SystemVerilog files and Verilate

In Step1:, the top module will have name of '__' because the module is parameterized.
protect-lib wrapper calls Verilated C++ via "DPI-C".

In Step2:, Verilator encodes symbols in protect-lib wrapper.
The encoded symbol includes "DPI-C" calling the Verilated C++ generated in Step1.
So the mismatch happens here.

This mismatch happens only on the Verilator in Step2. Other simulators don't have such trouble.

So I proposed
- Verilator in Step1 adds pragma in protect-lib wrapper
- Verilator in Step2 does not encode symbol if pragma is found

>As it does seem better to have the name be as close as possible to the original, that suggests track 1 - that is as output verilog should be as close as possible to input, when we write the wrapper it should use origName (or prettyName)?

I believe encoded/mangled name in protect-lib is inevitable to distinguish modules specialized by different paramete r values.

Reply:
I'm inclined to emitting protect-lib with encode-free name. (your suggestion). I will do some experiments.

Reply:
I experimentally added naming mechanism for parameterized hierarchical blocks that is encode free.
764f3a935c08dbce7d67b2117c74c5808646e965 and 764c5707857586f54811f3170a87d621b242d325

Reply:
tests look OK.
@wsnyder do you think the change (764f3a9 and 764c570) relaxes complexity ?